### ISSUE TITLE
ZZ/qall breaks "show your tabs and windows from last time"

### ISSUE THREAD

**Issue by mMontu** [cmcaine/tridactyl#350]
If you have multiple Firefox windows open and leave through <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Q</kbd>, all the windows are restored when Firefox restarts (due to `show your tabs and windows from last time` preference).

But if you leave through <kbd>Z</kbd><kbd>Z</kbd> then a single window is restore on the next session; all the other windows are lost.

**Comment by thejukra** [comment]
I can confirm this. Behaves just as mMontu described.

Tough after leaving with ZZ and restarting Firefox with Tridactyl the last opened windows may be restored from Library -> History -> Recently closed windows.

**Comment by bovine3dom** [comment]
Yeah, sorry about this. Last time I looked there wasn't a "quit Firefox" command, so we just have to close all of the windows. I'll have a look at it again - thanks for reporting it.

**Comment by bovine3dom** [comment]
We can probably just hack our away around this with a `TriStart` autocmd that checks to see if your most-recently-closed object was a window and then restores it.

**Comment by bovine3dom** [comment]
Dunno why I didn't think of this earlier: `bind ZZ !s killall firefox` works fine if you have the native messenger and killall installed.

**Comment by zeorin** [comment]
My problem is that when running `:qall`, some of the closed windows are not restored. Neither `CTRL`+`SHIFT`+`N` (native undo close window) nor `U` will recover the missing windows.
I currently have 7 windows open, with many tabs in each (one has 390 tabs, I know I have a problem).
When I last ran `:qall` (by accident, because, you know, vim-muscle-memory), and then opened Firefox again, only 3 of these windows were restored. The other windows are totally lost. If your’re like me, and keeping tabs around is how you keep your to-read lists, then this is a major problem.

It's not the 1st time that I've made this mistake.

# Recovering from this mistake

Close Firefox.

For anybody else that has the same problem, I (mostly) recovered from this by copying the missing windows from the last upgrade backup session.
In `$profile/sessionstore-backups` there are usually the following files:

```
previous.jsonlz4
recovery.baklz4
recovery.jsonlz4
upgrade.jsonlz4-$timestamp
```
There may be multiple of the upgrade backup files.

Your current session file will be at `$profile/sessionstore.jsonlz4`.

These files are compressed using a Mozilla-specific lz4 compression algo (don't ask). I use https://github.com/jusw85/mozlz4 to uncompress these.

In my experience, the `previous.jsonlz4`, `recovery.baklz4`, and `recovery.jsonlz4` files are basically the same as one another and as `sessionstore.jsonlz4` (there might be slight timestamp differences in the content data).

So these are of no help, either. The most recent upgrade session backup is where I look for the missing windows. The uncompressed files will be huge single-line JSON files. Hopefully your editor can pretty-print them, if not, you can use the `python -m json.tool` command-line utility (I run `:%!python -m json.tool` in Vim) to format them. The data structure is pretty clear, IMO, so you should be able to figure out what to copy where.

Then just re-compress the `sessionstore.jsonlz4` and place it in the root of your Firefox profile folder. Start Firefox, and presto!, your windows are back (assuming that the last upgrade backup was recent enough for you)

# Avoiding this mistake again

`:alias qall noop`

This will also handle `ZZ` and `:qa`, since they both call `:qall`.

**Comment by ironreality** [comment]
@bovine3dom Hi Oliver!
I have FF 92.0 and Tridactyl 1.21.1
In my case 'ZZ' typing closes FF without saving my tabs.
Also I tried to add `bind ZZ !s killall firefox` to ~/.tridactylrc
Could you elaborate more on how to use the workaround?

**Comment by bovine3dom** [comment]
On about:preferences is Firefox set to "show windows and tabs from last time"? Does `:viewconfig --user` show your ZZ bind?

**Comment by ironreality** [comment]
> :viewconfig --user

@bovine3dom Thank you for your fast response!
I added that bind manually right from FF window and now ':viewconfig --user' outputs the next:

```
{
  "configversion": "2.0",
  "hintfiltermode": "vimperator-reflow",
  "hintnames": "numeric",
  "nmaps": {
    "<": "back",
    ">": "forward",
    "ZZ": "!s killall firefox"
  },
  "update": {
    "lastchecktime": 1597518176214,
    "lastnaggedversion": "1.20.0"
  },
  "updatenaglastversion": "1.20.4"
}
```

But now pressing 'ZZ' does nothing...
In the same time issuing 'killall firefox' in my terminal closes FF.

**Comment by bovine3dom** [comment]
You need `:native` installed and working : )

**Comment by Wolfbeta** [comment]
> Dunno why I didn't think of this earlier: `bind ZZ !s killall firefox` works fine if you have the native messenger and killall installed.
> 
> Edit: this issue is so old that there's also the `U` key bound to `undo window` that will restore a lost window.

Safer still:

`bind ZZ !s osascript -e 'tell application "Firefox" to quit'`

This avoids sending SIGTERM directly and allows Firefox to perform its normal shutdown housekeeping. With killall, I noticed that closing a tab and immediately pressing ZZ could cause the tab to reappear after restart. Firefox had not yet written the updated session state to disk because of the session-store write interval. Quitting through AppleScript fixes that race.
